### PR TITLE
Remove GPU usage from our pipeline

### PR DESCRIPTION
**Pull request by Dashing-Nelson**
| Instance Type              | Shard-Fold Node Completion Time | Overall Pipeline Completion Time | CPU Utilisation | Accuracy |
|-----------------------------|----------------------------------|----------------------------------|-----------------|-----------|
| GPU                         | 1h13m                           | 5h21m                           | ~2%              | 0         |
| CPU with Parallel trial      | 1h41m                           | 2h38m                           | ~95%            | -1%       |
| CPU with sequential trial    | 6h36m                           | 12h18m                          | ~2%              | 0         |

# TL;DR

## Decision:
Make the modelling/tuning stage CPU‑first on high‑CPU Spot; remove GPU requests/limits (PR #1869). Keep GPU auto‑detection in Spark for future use.

## Why:
GPUs are a hard global bottleneck (16 cap; ~5–6 typically available). CPU Spot is abundant, cheaper, and XGBoost parallelizes well on CPUs.

## How:
Adaptive parallelism—evaluate a small batch of configs in parallel (≈2–4) and allocate threads per model from available CPUs (keep n_jobs=-1 in config, cap inner threads at runtime) to avoid over‑subscription.

## Impact:
Eliminated GPU scheduling waits; ~50% faster Shard×Fold node time and ~70% faster end‑to‑end pipeline in reps; all on CPU Spot. Docker image shrank 6.3 GB → 1.5 GB by dropping GPU deps.

## Trade‑offs:
Single fits may be slower than large GPU runs; Spot preemption risk (mitigated via retries and batch sizing); tuner/CPU allocation logic must evolve with cluster size.

## Ops guardrails:
Constrain pipeline concurrency so that Shard × Fold fan-out fits Spot CPU headroom; queuing moves to the CPU pool, not the global GPU gate.

## Metrics to watch:

Pipeline p50/p90, node wait time, CPU utilisation & Spot preemption rate, cost per pipeline, model quality vs baselines.

## Backout:

Re‑enable prior GPU requests/limits and reduce Shard×Fold to 3×3; Spark GPU auto‑config remains for fast revert.

# Docker image size has been reduced to 1.5 GB from 6.3 GB.

Screenshot 2025-10-29 at 16 48 27

# Performance Gain for Shard-X-Fold-X node

# Accuracy sheet (main vs this branch)
https://docs.google.com/spreadsheets/d/19SpxvToiHVm7zHpIDqr4gjTvBPRnfdnzqPgVpaIHW1s/edit?usp=sharing

# This branch Run:
https://argo.platform.dev.everycure.org/workflows/argo-workflows/mm-ts-sm-multi-trial-96f77018?tab=workflow&nodeId=mm-ts-sm-multi-trial-96f77018&uid=e26ee6c7-1271-4d87-add1-2d5e6c69ec98

# Main Branch Run:
https://argo.platform.dev.everycure.org/workflows/argo-workflows/mm-ts-sm-main-branch-747a94e0?tab=workflow&nodeId=mm-ts-sm-main-branch-747a94e0&uid=b5990de8-382b-4763-81c8-34344b577237

### Time taken
Screenshot 2025-10-09 at 15 01 08
## CPU Use
Screenshot 2025-10-16 at 13 19 19

### Time Taken
image
### CPU Use

Screenshot 2025-10-16 at 13 26 49

# Time taken for modelling run to complete (Over the weekend)

## With GPU
image

## With CPU
image

### Performance Gain
50% performance gain for node completion. Slashed running time to almost half. 1 hour overall change.
70% performance gain for the overall pipeline run. Thanks to the abundance of CPU.
90% Docker Image size reduction

### All of it running on Spot instances

Screenshot 2025-10-17 at 10 21 10

### NO GPU!!!!!!!!!!

# Description of the changes 

This pull request introduces several infrastructure and documentation improvements focused on supporting large-scale machine learning workloads, optimizing resource allocation, and adding robust GPU detection and configuration for Spark jobs. The most significant changes are the addition of new high-memory node pools for ML workloads, improved CPU/GPU allocation strategies, and enhancements to Spark resource management for GPU-enabled clusters.

**Infrastructure: Node Pool Expansion & Optimization**
- Added new `h3-standard-88` node pools (regular and spot) for large ML workloads, including appropriate taints and zone placement for high-availability and scheduling control. Updated node pool aggregation logic to include these new pools. [[1]](diffhunk://#diff-c8752c8cb6721b71f98987ba4a697feccb364bde13760069bdae5d6b436cb40fR119-R150) [[2]](diffhunk://#diff-c8752c8cb6721b71f98987ba4a697feccb364bde13760069bdae5d6b436cb40fL122-R163) [[3]](diffhunk://#diff-c8752c8cb6721b71f98987ba4a697feccb364bde13760069bdae5d6b436cb40fL131-R174) [[4]](diffhunk://#diff-c8752c8cb6721b71f98987ba4a697feccb364bde13760069bdae5d6b436cb40fR247-R275)
- Improved zone configuration for node pools: GPU pools now span multiple zones for reliability, and H3 pools are placed in a dedicated zone. Added `location_policy = "ANY"` to all pools for flexible scheduling. [[1]](diffhunk://#diff-c8752c8cb6721b71f98987ba4a697feccb364bde13760069bdae5d6b436cb40fR6-R7) [[2]](diffhunk://#diff-c8752c8cb6721b71f98987ba4a697feccb364bde13760069bdae5d6b436cb40fL15-R33) [[3]](diffhunk://#diff-c8752c8cb6721b71f98987ba4a697feccb364bde13760069bdae5d6b436cb40fR43) [[4]](diffhunk://#diff-c8752c8cb6721b71f98987ba4a697feccb364bde13760069bdae5d6b436cb40fL48-R52) [[5]](diffhunk://#diff-c8752c8cb6721b71f98987ba4a697feccb364bde13760069bdae5d6b436cb40fR61) [[6]](diffhunk://#diff-c8752c8cb6721b71f98987ba4a697feccb364bde13760069bdae5d6b436cb40fR78) [[7]](diffhunk://#diff-c8752c8cb6721b71f98987ba4a697feccb364bde13760069bdae5d6b436cb40fL81-R89) [[8]](diffhunk://#diff-c8752c8cb6721b71f98987ba4a697feccb364bde13760069bdae5d6b436cb40fR100)

**Documentation: Resource Allocation & Tuning**
- Added detailed documentation on intelligent CPU allocation for XGBoost, explaining how the tuner automatically balances parallel hyperparameter search and model training across available CPUs, preventing resource contention and maximizing efficiency.
- Documented the new parallel hyperparameter optimization implementation, describing the switch to scikit-optimize's `Optimizer` with ask/tell, adaptive parallelization, and the rationale for limiting parallel trials to maintain Bayesian optimization quality.

**Spark GPU Detection & Configuration**
- Added a robust `detect_gpus()` utility function to `spark_utils.py` that detects available GPUs using CuPy, environment variables, or `nvidia-smi`, with logging for troubleshooting.
- Updated Spark initialization to automatically configure GPU resources if GPUs are detected, including optimized settings for task concurrency and Rapids acceleration. Falls back to CPU-only configuration when no GPUs are found.

**Miscellaneous**
- Removed a redundant comment from `spark_utils.py` for code clarity.

## Fixes / Resolves the following issues:

- GPU scarcity. At max, 6 GPUs can be allocated as the demand for GPUs is very high.
- No spot instances for GPU. 
- takes almost 6 hours.

# Checklist:

- [X] Added label to PR (e.g. `enhancement` or `bug`)
- [X] Ensured the PR is named descriptively. FYI: This name is used as part of our changelog & release notes.
- [X] Looked at the diff on github to make sure no unwanted files have been committed. 
- [X] Made corresponding changes to the documentation
- [X] Added tests that prove my fix is effective or that my feature works
- [X] Any dependent changes have been merged and published in downstream modules
- [X] If breaking changes occur or you need everyone to run a command locally after
    pulling in latest main, uncomment the below "Merge Notification" section and
    describe steps necessary for people
- [X] Ran on sample data using `kedro run -e sample -p test_sample` (see [sample environment guide](https://docs.dev.everycure.org/onboarding/sample_environment/))

### CODE REVIEW

File: docs/src/infrastructure/TUNING_PARALLEL_OPTIMIZATION.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
+# Parallel Hyperparameter Optimization Implementation
```
Comment (pascalwhoop) [inline docs/src/infrastructure/TUNING_PARALLEL_OPTIMIZATION.md]: same here this is an update, not a docs page. if you want to you can add it to a release/changelog page

File: pipelines/matrix/src/matrix/pipelines/modelling/tuning.py [hunk 2]
```diff
@@ -101,22 +109,64 @@ def evaluate_model(**params):
...
+
+        # Run optimization with parallel evaluation
+        all_results = []
+        for _ in range(self.n_calls):
+            # Ask for next point(s) to evaluate
+            if self.n_parallel_evals == 1:
+                # Sequential: ask for one point at a time
+                next_x = optimizer.ask()
+                next_y = evaluate_model(next_x)
+                result = optimizer.tell(next_x, next_y)
```
[the remark this answers is not in the document]
Comment (JacquesVergine) [reply pipelines/matrix/src/matrix/pipelines/modelling/tuning.py]: Actually you do this on every iteration

File: pipelines/matrix/tests/pipelines/test_modelling.py [hunk 3]
```diff
@@ -534,7 +534,7 @@ def test_tune_parameters_regex_features_other_convention(
 def test_tune_parameters_train_test_split(tune_data: pd.DataFrame, grid_search_tuner: GridSearchCV) -> None:
     """Test proper handling of train/test splits."""
     mock_tuner = Mock()
-    mock_tuner._estimator = LogisticRegression()
+    mock_tuner.estimator = LogisticRegression()
```
Comment (JacquesVergine) [inline pipelines/matrix/tests/pipelines/test_modelling.py]: Nitpick: in this PR you have renamed class attributes to remove the leading underscore. Why?

I am asking for information, I have a slight opinion on the matter but I'm willing to be convinced :)

File: pipelines/matrix/tests/pipelines/test_modelling_tuning.py [hunk 4]
```diff
@@ -56,3 +59,57 @@ def test_nop_tuner_repr() -> None:
...
+
+        def fit(self, X, y=None):
+            return self
+
+        def predict(self, X):
+            # Simple prediction based on param1
+            return np.full(len(X), self.param1 % 2)
+
+    # Create synthetic data
+    np.random.seed(42)
```
Comment (JacquesVergine) [inline pipelines/matrix/tests/pipelines/test_modelling_tuning.py]: Nice

**Review by JacquesVergine — COMMENTED** [review]
Great work Nelson! I've asked some questions for my curiosity

File: pipelines/matrix/src/matrix/pipelines/modelling/tuning.py [hunk 5]
```diff
@@ -101,22 +109,64 @@ def evaluate_model(**params):
...
+            random_state=self.estimator.random_state if hasattr(self.estimator, "random_state") else None,
+        )
+
+        # Run optimization with parallel evaluation
+        all_results = []
+        for _ in range(self.n_calls):
+            # Ask for next point(s) to evaluate
+            if self.n_parallel_evals == 1:
+                # Sequential: ask for one point at a time
+                next_x = optimizer.ask()
```
[the remark this answers is not in the document]
Comment (Dashing-Nelson) [reply pipelines/matrix/src/matrix/pipelines/modelling/tuning.py]: so when n_parallel_evals = 1, we just make it sequential.

**Review by pascalwhoop — APPROVED** [review]
